src/sql/workbench/contrib/notebook/browser/notebookViews/insertCellsModal.ts

```diff
@@ -0,0 +1,326 @@
+import { ILogService } from 'vs/platform/log/common/log';
+import { IColorTheme, ICssStyleCollector, IThemeService, registerThemingParticipant } from 'vs/platform/theme/common/themeService';
+import * as DOM from 'vs/base/browser/dom';
+import { attachCheckboxStyler } from 'sql/platform/theme/common/styler';
+import { ServiceOptionType } from 'sql/platform/connection/common/interfaces';
+import { ServiceOption } from 'azdata';
+import * as DialogHelper from 'sql/workbench/browser/modal/dialogHelper';
+import { TextCellComponent } from 'sql/workbench/contrib/notebook/browser/cellViews/textCell.component';
+import { ComponentFactoryResolver, ViewContainerRef } from '@angular/core';
+import { ICellModel } from 'sql/workbench/services/notebook/browser/models/modelInterfaces';
+import { NotebookModel } from 'sql/workbench/services/notebook/browser/models/notebookModel';
+import { inputBorder, inputValidationInfoBorder, editorBackground } from 'vs/platform/theme/common/colorRegistry';
+import { localize } from 'vs/nls';
+import { NotebookViewsExtension } from 'sql/workbench/services/notebook/browser/notebookViews/notebookViewsExtension';
+import { attachButtonStyler } from 'vs/platform/theme/common/styler';
+import { toJpeg } from 'html-to-image';
+
+export class CellOptionsModel {
+	private _optionsMap: { [name: string]: any } = {};
+
+	constructor(
+		optionsMetadata: ServiceOption[],
+		private onInsert: (cell: ICellModel) => any,
+		private _context: NotebookViewsExtension,
+	) {
+		optionsMetadata.forEach(optionMetadata => {
+			let defaultValue = this.getDisplayValue(optionMetadata, optionMetadata.defaultValue);
+			this._optionsMap[optionMetadata.name] = {
+				optionMetadata: optionMetadata,
+				defaultValue: defaultValue,
+				currentValue: defaultValue
+			};
+		});
+	}
+
+	private getDisplayValue(optionMetadata: ServiceOption, optionValue: any): any {
+		let displayValue: any;
+		switch (optionMetadata.valueType) {
+			case ServiceOptionType.boolean:
+				displayValue = DialogHelper.getBooleanValueFromStringOrBoolean(optionValue);
+				break;
+			case ServiceOptionType.category:
+				let optionName = optionValue;
+				if (!optionName && optionMetadata.categoryValues[0]) {
+					optionName = optionMetadata.categoryValues[0].name;
+				}
+				displayValue = DialogHelper.getCategoryDisplayName(optionMetadata.categoryValues, optionName);
+				break;
+			case ServiceOptionType.string:
+				displayValue = optionValue ? optionValue : '';
+		}
+		return displayValue;
+	}
+
+	restoreCells() {
+		for (let key in this._optionsMap) {
+			let optionElement = this._optionsMap[key];
+			if (optionElement.currentValue === true) {
+				const activeView = this._context.getActiveView();
+				const cellToInsert = activeView.getCell(optionElement.optionMetadata.name);
+				if (cellToInsert) {
+					this.onInsert(cellToInsert);
+				}
+			}
+		}
+	}
+
+	public setOptionValue(optionName: string, value: any): void {
+		if (this._optionsMap[optionName]) {
```

Comment:
Why are you only setting it if it's there? This also means if the value is falsey (empty string/0) we won't set it

Reply:
Well, the check is there to make sure that an entry been created in optionsMap with that key so that we are not trying to assign a value to undefined. Value can be true or false, we just need to make sure we're assigning it to a valid object. 

I think that is more obvious now that I have assigned a type definition to optionsMap.

Reply:
Currently this is only going to do that for truthy values, so false values won't be set. Which sounds like that would be unintended. You need to specifically check for undefined if that's what you're trying to here.

Reply:
Do you mean something like this? I was under the impression that _this._optionsMap[optionName]_ would be false if it was undefined, and true if there was an object at that key.

public setOptionValue(optionName: string, value: boolean): void {
		if (this._optionsMap[optionName] !== undefined) {

Reply:
Boolean logic in javascript is done using truthiness by default https://developer.mozilla.org/en-US/docs/Glossary/Falsy

So yes, to check that a value is specifically undefined and not false you would need to check that it !== undefined.